pandas/core/dtypes/astype.py

```diff
@@ -145,11 +145,18 @@ def astype_nansafe(
         if is_datetime64_dtype(dtype):
             from pandas import to_datetime
 
+            datetime_values = to_datetime(arr.ravel()).values.reshape(arr.shape)
+
+            if dtype == "<M8[D]":
+                datetime_values = np.datetime_as_string(datetime_values, unit="D")
+                datetime_values = datetime_values.astype(dtype)
+
             return astype_nansafe(
```

Comment:
i think this recursive astype_nansafe call might be unnecessary?  if we just get rid of that, does that solve the problem?

Reply:
Hi If I get rid of astype_nansafe call, it will fail for type `datetime64[D]` . The  integer overflow still occur  for type `datetime64[D]`
```
@pytest.mark.parametrize(
    "datetime64",
    ["datetime64", "datetime64[s]", "datetime64[m]", "datetime64[h]", "datetime64[D]"],
)
def test_integer_overflow_datetime64(datetime64):
    # GH 28045

    result = Series(["9/22/1677", "4/11/2262"], dtype="U").astype(datetime64)
    expected = Series(["9/22/1677", "4/11/2262"], dtype="datetime64[ns]")

    tm.assert_series_equal(result, expected)

```